### PR TITLE
v6 - Card - Dual Brand Selection

### PR DESCRIPTION
**Pull request by tkuntubayev**
## Description
[//]: # (Include a short summary of your changes)
[//]: # (If this is a new feature: attach screenshots or a video if applicable)
[//]: # (If this is a bug fix: include a reproduction path)
Adds UI to allow selection of a brand for Dual Branded Cards with Selection and adds corresponding Analytics events.

[Recording](https://github.com/user-attachments/assets/fa2bd1bb-15f7-4657-936b-12a05b1153a6)

## Checklist 
- [x] Code is unit tested
- [x] Changes are tested manually

## Ticket Number
COSDK-1193

### CODE REVIEW

File: card/src/main/java/com/adyen/checkout/card/internal/ui/view/CardNumberField.kt [hunk 1]
```diff
@@ -138,13 +154,74 @@ private fun DetectedBrandsList(
...
+                                    shape = RoundedCornerShape(Dimensions.CornerRadius),
+                                )
+                                .background(color = CheckoutThemeProvider.colors.background)
+                        } else {
+                            Modifier
+                        },
+                    )
+                    .clickable { onBrandSelect(brandItem.brand) }
+                    .padding(Dimensions.Spacing.ExtraSmall)
+                    .clip(RoundedCornerShape(Dimensions.CornerRadius)),
```
[the remark this answers is not in the document]
Comment (tkuntubayev) [reply card/src/main/java/com/adyen/checkout/card/internal/ui/view/CardNumberField.kt:223]: The first clip is applied to the brand logo raw image, the second clip is to the combined image with paddings. Having just one will make either the raw image rectangular or paddings being rectangular which will mismatch the UI design.

File: card/src/main/java/com/adyen/checkout/card/internal/ui/CardComponent.kt [hunk 2]
```diff
@@ -244,6 +251,39 @@ constructor(
...
+
+                val event = DualBrandCardEvents.dualBrandSelectionDisplayed(
+                    component = paymentMethodType,
+                    selectedBrand = cardBrandState.shopperSelectedCardBrandData.cardBrand,
+                    brandOptions = cardBrandState.cardBrandDataList,
+                )
+                analyticsManager.trackEvent(event)
+            }
+            .launchIn(coroutineScope)
+    }
```
[the remark this answers is not in the document]
Comment (tkuntubayev) [reply card/src/main/java/com/adyen/checkout/card/internal/ui/CardComponent.kt]: The accessing `componentState.value` inside a reactive flow pipeline is replaced with flow operators

File: card/src/main/java/com/adyen/checkout/card/internal/ui/view/CardNumberField.kt [hunk 3]
```diff
@@ -138,13 +154,74 @@ private fun DetectedBrandsList(
...
             is CardBrandViewState.Placeholder -> BrandLogo(txVariant = null)
             is CardBrandViewState.SingleBrand -> BrandLogo(cardBrandViewState.brand.txVariant)
-            is CardBrandViewState.DualBrand -> cardBrandViewState.brands.forEach { brand ->
-                BrandLogo(brand.txVariant)
-            }
-            // TODO - Co-badged selectable cards [COSDK-1193]
-            is CardBrandViewState.SelectableDualBrand -> cardBrandViewState.brands.forEach { brandItem ->
-                BrandLogo(brandItem.brand.txVariant)
-            }
+            is CardBrandViewState.DualBrand -> DualBrandLogos(cardBrandViewState.brands)
```
Comment (jreij) [inline card/src/main/java/com/adyen/checkout/card/internal/ui/view/CardNumberField.kt:157]: Can you update the compose preview to showcase something related to dual branding?

File: card/src/main/java/com/adyen/checkout/card/internal/ui/view/CardNumberField.kt [hunk 4]
```diff
@@ -138,13 +154,74 @@ private fun DetectedBrandsList(
...
+                    .semantics {
+                        role = Role.RadioButton
+                        selected = brandItem.isSelected
+                    }
+                    .clip(RoundedCornerShape(Dimensions.CornerRadius))
+                    .then(
+                        if (brandItem.isSelected) {
+                            Modifier
+                                .border(
+                                    width = 1.dp,
```
Comment (jreij) [inline card/src/main/java/com/adyen/checkout/card/internal/ui/view/CardNumberField.kt:212]: I see this 1dp hardcoded here and in `CheckoutNetworkLogo`, both representing borders. Is this a dimension we should extract?

File: card/src/main/java/com/adyen/checkout/card/internal/ui/view/CardNumberField.kt [hunk 5]
```diff
@@ -218,7 +296,7 @@ private fun CardNumberFieldIcon(
                 )
             }
 
-            else -> DetectedBrandsList(cardBrandViewState)
+            else -> DetectedBrandsList(cardBrandViewState, onBrandSelect)
```
Comment (jreij) [inline card/src/main/java/com/adyen/checkout/card/internal/ui/view/CardNumberField.kt]: Not directly related to this PR but can we explicitly use `CardNumberTrailingIcon.BrandLogos` instead of `else` here?

Comment (tkuntubayev) [reply card/src/main/java/com/adyen/checkout/card/internal/ui/view/CardNumberField.kt]: For that to happen, we'd need to handle the `null` case now as `trailingIcon` is nullable. 
We could:
1. make it non-nullable by exiting early in `CardNumberFieldIcon`:
```val trailingIcon = state.trailingIcon as? CardNumberTrailingIcon ?: return```
2. or come up with a default icon to show for `null` case

 wdyt?

File: card/src/main/java/com/adyen/checkout/card/internal/ui/view/CardNumberField.kt (hunk 4 above)
Comment (tkuntubayev) [reply card/src/main/java/com/adyen/checkout/card/internal/ui/view/CardNumberField.kt:212]: The one in the `CheckoutNetworkLogo` represents a shadow blur radius, not really a border. Combining them under 1 dimension might be misleading. So far each of them is used only once. Let me know wdyt on whether we'd like to extract them under 2 dimensions, put under 1 common dimension or keep it as-is

File: card/src/main/java/com/adyen/checkout/card/internal/ui/CardComponent.kt [hunk 6]
```diff
@@ -245,6 +251,40 @@ constructor(
...
+            )
+            analyticsManager.trackEvent(event)
+        }
+        onIntent(intent)
+    }
+
+    private fun subscribeToDualBrandSelectionAppearAnalyticsEvents() {
+        componentState
+            .map { it.cardBrandState }
+            .distinctUntilChanged { old, new ->
```
Comment (jreij) [inline card/src/main/java/com/adyen/checkout/card/internal/ui/CardComponent.kt:275]: I struggled to understand this logic here and with quite a bit of help from AI, I now get it. I was sure we could re-write this in a more understandable way but I was wrong, this seems to be the best way.

However, this logic is very fragile to an accidental refactor. For example if someone tries to "optimize" by moving `mapNotNull` before `distinctUntilChanged` or making `distinctUntilChanged` only return if `isOldDualBrandWithSelection == false` and `isNewDualBrandWithSelection == true`, it would cause a bug.

I suggest we add a comment explaining the logic a little bit (and warning against changing it) and adding some unit tests to cover all the edge cases.

File: ui/src/main/java/com/adyen/checkout/ui/internal/element/input/CheckoutTextFieldDecorationBox.kt [hunk 7]
```diff
@@ -98,7 +98,7 @@ internal fun CheckoutTextFieldDecorationBox(
                 .styledBackground(style, isFocused, isError)
                 .fillMaxWidth()
                 .heightIn(Dimensions.MinTouchTarget)
-                .padding(horizontal = Dimensions.Spacing.Large, vertical = Dimensions.Spacing.Medium),
+                .padding(horizontal = Dimensions.Spacing.Large),
```
Comment (jreij) [inline ui/src/main/java/com/adyen/checkout/ui/internal/element/input/CheckoutTextFieldDecorationBox.kt:101]: Is this change intended?

File: ui/src/main/java/com/adyen/checkout/ui/internal/element/input/CheckoutTextFieldDefaults.kt [hunk 8]
```diff
@@ -18,7 +18,7 @@ internal object CheckoutTextFieldDefaults {
         attributes: InternalAttributes,
     ): InternalTextFieldStyle {
         return InternalTextFieldStyle(
-            backgroundColor = colors.container,
+            backgroundColor = colors.background,
```
Comment (jreij) [inline ui/src/main/java/com/adyen/checkout/ui/internal/element/input/CheckoutTextFieldDefaults.kt:21]: Is this change intended?

File: card/src/main/java/com/adyen/checkout/card/internal/ui/view/CardNumberField.kt (hunk 4 above)
Comment (jreij) [reply card/src/main/java/com/adyen/checkout/card/internal/ui/view/CardNumberField.kt:212]: Ah in that case let's just keep it as it is then

File: card/src/main/java/com/adyen/checkout/card/internal/ui/view/CardNumberField.kt (hunk 5 above)
Comment (jreij) [reply card/src/main/java/com/adyen/checkout/card/internal/ui/view/CardNumberField.kt]: Is null even a possible use case with this field? if not then maybe we can just not return anything and put a comment that null is not expected

File: ui/src/main/java/com/adyen/checkout/ui/internal/element/input/CheckoutTextFieldDefaults.kt (hunk 8 above)
Comment (tkuntubayev) [reply ui/src/main/java/com/adyen/checkout/ui/internal/element/input/CheckoutTextFieldDefaults.kt:21]: Yes, to show proper background for `TextInputField` and have contrast with the container color of Selectable Dual-Brand logos container. With this it gets aligned with the current Figma design

File: card/src/main/java/com/adyen/checkout/card/internal/analytics/DualBrandCardEvents.kt [hunk 9]
```diff
@@ -0,0 +1,38 @@
+/*
```
Comment (jreij) [inline card/src/main/java/com/adyen/checkout/card/internal/analytics/DualBrandCardEvents.kt:1]: All commits are missing the ticket number, not sure if the agent forgot to add it

File: ui/src/main/java/com/adyen/checkout/ui/internal/element/input/CheckoutTextFieldDecorationBox.kt (hunk 7 above)
Comment (tkuntubayev) [reply ui/src/main/java/com/adyen/checkout/ui/internal/element/input/CheckoutTextFieldDecorationBox.kt:101]: Yes, the vertical padding adds unnecessary space for the whole view when the content has it's own paddings like the container padding of selectable Dual-Brand logos. The height of the view is constrained by `.heightIn(Dimensions.MinTouchTarget)` and after inner talk, the vertical padding was suggested for removal.

File: card/src/main/java/com/adyen/checkout/card/internal/analytics/DualBrandCardEvents.kt (hunk 9 above)
Comment (tkuntubayev) [reply card/src/main/java/com/adyen/checkout/card/internal/analytics/DualBrandCardEvents.kt:1]: The agent indeed didn't pick up the commit instructions part of AGENTS.md, so I did commit myself and wasn't aware of the ticket number need to be in each commit.

File: card/src/main/java/com/adyen/checkout/card/internal/ui/view/CardNumberField.kt (hunk 3 above)
Comment (tkuntubayev) [reply card/src/main/java/com/adyen/checkout/card/internal/ui/view/CardNumberField.kt:157]: Added previews for both types of dual branding

File: card/src/main/java/com/adyen/checkout/card/internal/ui/CardComponent.kt (hunk 6 above)
Comment (tkuntubayev) [reply card/src/main/java/com/adyen/checkout/card/internal/ui/CardComponent.kt:275]: Sounds good, I've extended Unit tests with more edge case scenarios. Also added supportive/warning comments, let me know how this looks now

File: card/src/main/java/com/adyen/checkout/card/internal/analytics/DualBrandCardEvents.kt (hunk 9 above)
Comment (jreij) [reply card/src/main/java/com/adyen/checkout/card/internal/analytics/DualBrandCardEvents.kt:1]: Yeah, we include the ticket number in the commit to make it easier for someone doing git blame / history in the future to quickly look up the ticket that introduced a certain change

Comment (jreij) [reply card/src/main/java/com/adyen/checkout/card/internal/analytics/DualBrandCardEvents.kt:1]: Maybe you can ask an agent to rebase all the commits and just add the ticket number to the message?

Comment (tkuntubayev) [reply card/src/main/java/com/adyen/checkout/card/internal/analytics/DualBrandCardEvents.kt:1]: Rebased, all commits now have the ticket number

File: card/src/main/java/com/adyen/checkout/card/internal/ui/view/CardNumberField.kt (hunk 5 above)
Comment (tkuntubayev) [reply card/src/main/java/com/adyen/checkout/card/internal/ui/view/CardNumberField.kt]: Currently this field is always populated by `CardViewStateProducer` with the expected `TrailingIcon` subtype so in practice it should never be `null`. The `else` is replaced with proper subtype check and early exit with supportive comment added

**Review by jreij — APPROVED** [review]
🚀
